nix/codex.nix

```
@@ -0,0 +1,380 @@
+        };
+
+        rewardRecipient = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = "Address to send payouts to (eg rewards and refunds).";
+        };
+      };
+
+      prover = {
+        enable = mkEnableOption "Enable the 'persistence prover' subcommand.";
+
+        circuitDir = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = "Directory where Codex will store proof circuit data.";
+        };
+
+        circomR1cs = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = "The r1cs file for the storage circuit.";
+        };
+
+        circomWasm = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = "The wasm file for the storage circuit.";
+        };
+
+        circomZkey = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = "The zkey file for the storage circuit.";
+        };
+
+        circomNoZkey = mkEnableOption "Ignore the zkey file - use only for testing!";
+
+        proofSamples = mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = "Number of samples to prove.";
+        };
+
+        maxSlotDepth = mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = "The maximum depth of the slot tree.";
+        };
+
+        maxDatasetDepth = mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = "The maximum depth of the dataset tree.";
+        };
+
+        maxBlockDepth = mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = "The maximum depth of the network block merkle tree.";
+        };
+
+        maxCellElements = mkOption {
+          type = types.nullOr types.int;
+          default = null;
+          description = "The maximum number of elements in a cell.";
+        };
+      };
+
+      extraArgs = mkOption {
+        type = types.nullOr (types.listOf types.str);
+        default = null;
+        description = "Additional arguments to pass to the Codex binary.";
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users.users = optionalAttrs (cfg.service.user == "codex") {
+      codex = {
+        group = cfg.service.group;
+        home = cfg.dataDir;
+        description = "Codex service user";
+        isSystemUser = true;
+      };
+    };
+
+    users.groups = optionalAttrs (cfg.service.user == "codex") {
+      codex = { };
+    };
+
+    systemd.tmpfiles.rules = [
+      "d ${cfg.dataDir or "/var/lib/codex"} 0755 codex codex"
+    ];
+
+    systemd.services.codex = {
+      description = "Codex Node";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        User = cfg.service.user;
+        Group = cfg.service.group;
```

Why not just use a `DynamicUser` and skip user creation altogether?
```suggestion
        DynamicUser = true;
```
We could also add some hardening options:
```
        PrivateTmp = true;
        ProtectHome = true;
        ProtectSystem = "full";
        NoNewPrivileges = true;
        PrivateDevices = true;
        MemoryDenyWriteExecute = true;
```